src/core/lwt.ml

```diff
@@ -818,6 +823,25 @@ struct
     with exn ->
       current_storage := saved_storage;
       raise exn
+
+  type setup_teardown = unit -> (unit -> unit)
```

Comment:
I realize this is how `with_key` is done, to get around the value restriction, but I prefer the user being oblivious to that technicality, and that the user-facing type be more natural. Does the following code work without imposing any additional overhead?
```ocaml
type 'a context = {
  setup : unit -> 'a;
  teardown : 'a -> unit;
}

type private_context = unit -> unit -> unit
  
let cur_context : private_context option ref = ref None
    
let to_private {setup;teardown} = fun () -> let x = setup () in fun () -> teardown x
  
let with_context c f =
  let saved = !cur_context in
  let setup_teardown = to_private c in
  let _ = cur_context := Some setup_teardown in
  let teardown = setup_teardown () in
  try
    let result = f () in
    let _ = teardown () in
    let _ = cur_context := saved in
    result
  with exn ->
    let _ = teardown () in
    let _ = cur_context := saved in
    raise exn
```

Also wondering whether it makes sense to have:
```ocaml
type 'a context = {
  setup : unit -> 'a;
  teardown : 'a -> unit;
  teardown_exn : 'a -> exn -> unit;
}
(* or *)
type ('a,'b) context = {
  setup : unit -> 'a;
  teardown : 'b state -> 'a -> unit;
}
```
To allow to teardown differently based on whether promise resolved or not.

Reply:
Also, do we need either special implementation or a new `with_context` like function for `Lwt_stream.t` or will that just work?

Reply:
Interface updated to 

```ocaml
val with_setup_teardown : setup:(unit -> 'a) -> teardown:('a -> unit) -> (unit -> 'c) -> 'c
```

I will review your comment about Lwt_stream.t

Reply:
I tested and if the stream map etc are done within a setup_teardown, the same functions are carried for streams